### PR TITLE
Disconnect List Blocking and Tracker Network Summary

### PR DESCRIPTION
**Pull request by subsymbolic**
Asana Issue URL: 
Disconnect List Blocking: https://app.asana.com/0/414235014887631/471718733961916
Tracker Networks Summary: https://app.asana.com/0/414235014887631/364367284740782

## Description
• Downloads the disconnect list from the server
• Adds disconnect list tracker blocking
• Adds a skeleton privacy grade page
• Displays tracker network count (currently static, live updates will be added in future tasks)

NOTE: Data sync service and storage and live are covered by separate tasks

## Steps to Test this PR:
Test One:
1. Navigate to duckduckgo.com
1. Open the Privacy Dashboard
1. Once loading has finished, observe that the major tracker number is 0

Test two:
1. Navigate to evanscycles.com
1. Open the Privacy Dashboard
1. Once loading has finished, observe that the major tracker number is 6

### CODE REVIEW

**Comment by CDRussell** [comment]
Noticing a performance slow down when viewing the privacy dashboard from evanscycles.com. It is taking anywhere up to 5s to display the privacy dashboard activity.

Agreed with @subsymbolic to tackle in another ticket.

**Comment by CDRussell** [comment]
Crash noticed when viewing evanscycles.com due to 

    11-27 13:26:00.763 W/System.err: java.util.ConcurrentModificationException
    11-27 13:26:00.763 W/System.err:     at java.util.ArrayList$Itr.next(ArrayList.java:831)
    11-27 13:26:00.763 W/System.err:     at com.duckduckgo.app.trackerdetection.TrackerDetector.shouldBlock(TrackerDetector.kt:38)
    11-27 13:26:00.763 W/System.err:     at com.duckduckgo.app.browser.BrowserWebViewClient.block(BrowserWebViewClient.kt:84)
    11-27 13:26:00.763 W/System.err:     at com.duckduckgo.app.browser.BrowserWebViewClient.shouldInterceptRequest(BrowserWebViewClient.kt:74)
    11-27 13:26:00.763 W/System.err:     at com.android.webview.chromium.Ap.shouldInterceptRequest(WebViewContentsClientAdapter.java:52)

File: app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt [hunk 1]
```diff
@@ -53,15 +58,15 @@ class BrowserWebViewClient @Inject constructor(private val requestRewriter: Duck
...
     override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
-        webViewClientListener?.loadingStateChange(true)
+        webViewClientListener?.loadingStarted()
         webViewClientListener?.urlChanged(url)
     }
 
     override fun onPageFinished(view: WebView?, url: String?) {
-        webViewClientListener?.loadingStateChange(false)
+        webViewClientListener?.loadingFinished()
+        webViewClientListener?.urlChanged(url)
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt]: why is this needed?

File: app/src/main/java/com/duckduckgo/app/global/Url.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.duckduckgo.app.global
+
+import android.net.Uri
+import android.net.Uri.parse
+import com.duckduckgo.app.global.UrlScheme.Companion.http
+
+class UrlScheme {
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/global/Url.kt]: Don't think we need this class here, do we? could you define those http/https constants at the top level of this file (but private?) ?

File: app/src/main/java/com/duckduckgo/app/global/Url.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+fun Uri.withScheme(): Uri {
+    if (scheme == null) {
+        return parse("${http}://${toString()}")
+    }
+
+    return this
+}
+
+/**
+ * Trys to resolve a host (even if the scheme is missing), returning
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/global/Url.kt]: typo: `Trys`

File: app/src/main/java/com/duckduckgo/app/global/Url.kt (hunk 2 above)
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/global/Url.kt]: Nice. Although can i suggest we name the file to contain `Extension`. I've went for `ViewExtension`, `EditTextExtension` etc...

`UriExtension` perhaps?

File: app/src/main/java/com/duckduckgo/app/privacydashboard/PrivacyDashboardActivity.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            android.R.id.home -> {
+                super.onBackPressed()
+                return true
+            }
+        }
+        return super.onOptionsItemSelected(item)
+    }
+
+    private fun render(siteMonitor: SiteMonitor) {
+        domain.text = Uri.parse(siteMonitor.url).host
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/privacydashboard/PrivacyDashboardActivity.kt]: should some of this be done in the `ViewModel` instead?

File: app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDetector.kt [hunk 5]
```diff
@@ -16,25 +16,26 @@
...
     }
 
     fun hasClient(name: ClientName): Boolean {
-        return clients.filter { it.name == name }.isNotEmpty()
+        return clients.any { it.name == name }
     }
 
     fun shouldBlock(url: String, documentUrl: String, resourceType: ResourceType): Boolean {
-        for (client: TrackerDetectionClient in clients) {
+        for (client: Client in clients) {
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDetector.kt]: why not use `clients.any` as you've done above?

File: app/src/main/java/com/duckduckgo/app/trackerdetection/model/NetworkTrackers.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    fun network(url: String): String? {
+
+        val lookupHost = Uri.parse(url)?.baseHost() ?: return null
+
+        return trackers
+                .filter { it: NetworkTracker ->
+                    val trackerHost = Uri.parse(it.url).baseHost()
+                    val networkHost = Uri.parse(it.networkUrl).baseHost()
+                    lookupHost == trackerHost || lookupHost == networkHost || lookupHost.endsWith(".$trackerHost") || lookupHost.endsWith(".$networkHost")
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/trackerdetection/model/NetworkTrackers.kt]: i'm finding this line hard to parse... pull into its own method - the name of which can explain what it does?

File: app/build.gradle [hunk 7]
```diff
@@ -85,13 +86,14 @@ dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jre7:$kotlin_version"
     implementation "com.android.support:appcompat-v7:$supportLibrary"
     implementation "com.android.support:design:$supportLibrary"
-    implementation 'com.android.support.constraint:constraint-layout:1.0.2'
-    implementation 'com.squareup.okhttp3:okhttp:3.9.0'
-    implementation 'com.squareup.retrofit2:retrofit:2.3.0'
-    implementation 'com.squareup.retrofit2:adapter-rxjava2:2.3.0'
+    implementation "com.android.support.constraint:constraint-layout:1.0.2"
```
[the remark this answers is not in the document]
Comment (subsymbolic) [reply app/build.gradle]: I did as the single quotes don't work with inner variables

File: app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt (hunk 1 above)
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt]: The url is often different by the end of a page load

File: app/src/main/java/com/duckduckgo/app/global/Url.kt (hunk 2 above)
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/global/Url.kt]: They're used in a few places throughout the app - we often want about http/https. I can move them into a separate file though

File: app/src/main/java/com/duckduckgo/app/global/Url.kt (hunk 3 above)
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/global/Url.kt]: Yikes!

File: app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDetector.kt (hunk 5 above)
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDetector.kt]: Good question, probably just kotlin newbie conservatism. There is a lot more happening here and I thought it was more obvious as is. But I'll take the plunge and update this too!

File: app/src/main/java/com/duckduckgo/app/global/Url.kt (hunk 3 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/global/Url.kt]: I'm just gonna assume you've been watching a lot of rugby. 😀

File: app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt (hunk 1 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt]: But wouldn't you have already seen a call to `onPageStarted` whenever the URL changes?

If you have seen this happening, and the need for this call, then 👍 . but seems a bit wrong if we haven't seen that happening.

File: app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDetector.kt (hunk 5 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDetector.kt]: I don't mind which approach you go for, but feel like we use the same approach in the same class.

File: app/src/main/java/com/duckduckgo/app/privacydashboard/PrivacyDashboardActivity.kt (hunk 4 above)
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/privacydashboard/PrivacyDashboardActivity.kt]: Agreed, this is getting ripped out with the live data update, mind if I change it there?

**Comment by CDRussell** [comment]
Approved, assuming there is only minimal work needed to resolve merge conflict
